docs/dev/contributing.rst

```diff
@@ -0,0 +1,78 @@
+Workflow
+========
+
+Recommended one-time setup
+--------------------------
+
+Remove ``--global`` if you don't want it to apply to other git repos on your computer: ::
+
+  git config --global user.name "Your Name"
+  git config --global user.email you@something.edu
+
+Other git config you might find useful: ::
+
+  git config --global push.default simple # makes git only push the current branch, which is useful for not accidentally messing things up
+  git config --global color.{ui,status,branch,diff,interactive} auto # makes various interfaces usefully colorful
+  git config --global log.decorate true # shows branch information in `git log` by default
+
+For normal, non-urgent features and bug fixes
+---------------------------------------------
+
+The following workflow applies if you've already been added as a collaborator to the repository.  If not, you should fork it using the button on Github, add it as a remote, and replace all of the ``git push`` steps with ``git push <remote name>``.
+
+From the directory ``/esp``: ::
+
+  git checkout main
+  git pull
+  ./update_deps.sh # on vagrant: see vagrant docs; no need to bother if deps haven’t changed
+  ./manage.py update # on vagrant: fab manage:cmd=update
+  git checkout -b new-branch-name
+
+Write some code!
+Test your code!
+
+Look at what you’ve changed (``git status`` and/or ``git diff``) ::
+
+  git commit -a -m "Write a commit message here"
```

Comment:
Is it best practice to recommend this? I know this is the easiest way to do this, but it's also easy to do the wrong thing with this.

Rather than recommend ``-a``, I'd rather recommend combinations of ``git status``, ``git add`` (or even ``git add -p``), and ``git diff --cached`` to make sure that the correct set of changes is included. Here ``git status`` is replacing the ``-a`` flag by making sure you don't forget any files.

I myself use ``-m``, but it does encourage either writing short useless commit messages, or way-too-long one-line commit message. Can we either omit the ``-m``, or also add ``-e`` (the ``-m`` / ``-e`` combination will set the first line of the commit to that message, but still open up the editor for editing the full commit message).

Can you also link to <http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html> and <https://github.com/blog/926-shiny-new-commit-styles>?